### PR TITLE
Additional front-end guidelines

### PR DESCRIPTION
**Pull request by stevenspiel**

### CODE REVIEW

**Comment by stevenspiel** [comment]
I made comments and clarified some things we talked about in WTL

**Comment by andycandrea** [comment]
:+1:

File: README.md [hunk 1]
```diff
@@ -142,14 +146,24 @@ Model.reset_column_information
 
 # JavaScript
 * Use CoffeeScript over JavaScript
+* Namespace javascript objects for related chuncks of funcationlity (Components, Engines, etc...)  [example implementation](samples/js_namespace.md)
```
Comment (mjs2600) [inline README.md]: There's a type in functionality

File: README.md [hunk 2]
```diff
@@ -142,14 +146,24 @@ Model.reset_column_information
...
+* Use js classes and encapsulation wherever possible ([class example](samples/js_class.md), [encapsulation example](samples/js_encapsulation.md))
 * Use a package manager like Bower for installing dependencies? (This item is questionable as some people raised issues with using Bower with Rails. We can discuss further.)
 * Prefer a JavaScript framework over vanilla JavaScript. (This means don't roll your own custom event library or other things that exist out there. But do make sure the company is on board with any new libraries that are used before including them in a project)
 * Use IIFE or Object Literal notation
 * Separate responsibilities of vanilla JavaScript into separate entities.
 * Query DOM elements using `js-` or `data-` attributes instead of CSS classes
 * Do not rely on or store data in the DOM. (This is particularly true from a security stand point. Do not put secure or sensitive information in the DOM. For example do not have a hidden field with a SSN, or that determines whether someone is an admin).
+* Coffeescript file extensions not include `.js` where possible. Prefer `foo.coffee` to `foo.js.coffee.`
```
Comment (mjs2600) [inline README.md]: Have we thought about switching to ES6 for new apps?

**Comment by mjs2600** [comment]
:+1:

File: README.md (hunk 2 above)
Comment (emerson-h) [inline README.md]: Which browsers support ES6, and is the standard finalized yet?

File: README.md (hunk 1 above)
Comment (emerson-h) [inline README.md]: I would really like to see a "Unit test JS objects wherever possible" added as the next point here

Comment (mjs2600) [inline README.md]: @emerson-h :+1:

File: README.md (hunk 2 above)
Comment (mjs2600) [inline README.md]: > is the standard finalized yet?
Yep: http://www.infoq.com/news/2015/06/ecmascript-2015-es6

> Which browsers support ES6
https://kangax.github.io/compat-table/es6/
I think for now you'd transpile down using babel and the asset pipeline (Like we do for coffeescript), but given that it's the standard, at some point that won't be necessary.

Comment (emerson-h) [inline README.md]: As long as it transpiles appropriately, I would say we should investigate ES6 and use native JS classes.  

It is worth noting that Chrome does not support ES6 classes at the moment.

Comment (mjs2600) [inline README.md]: @stevenspiel any thoughts?

**Comment by josephjaber** [comment]
Let's just use ES7 without a transpiler. If a customer can't execute it, we can just have them read the JS source to figure out what the code would have done. They can pretend that's what happened in their minds.

:shipit: :shipit:

**Comment by mjs2600** [comment]
@josephjaber That'll work on lynx too! :shipit:

File: README.md [hunk 3]
```diff
@@ -78,6 +78,10 @@ of development experience.
   [In Defense of Copy Paste](http://zacharyvoase.com/2013/02/08/copypasta/)
   for more information. While this is applicable in all forms of code, this
   is particularly problematic in view code.
+* Put data attributes in a hash when using a view helper e.g. `f.input :foo, data: { input_method: :foo }`. Note that underscores in keys will be translated to dashes when rendered to html. The previous example's data attribute will render: `data-input-method="foo"`.
+* Prefer `f.collection_select :foo` over `f.select :foo, options_from_collection_for_select`
+* Use rails helper for labels wherever possible
+* Ensure the `label_for` attribute connects with the correct form input
```
Comment (kathleenlan) [inline README.md]: Reword as `Ensure that label tags connect with the correct form input`

File: README.md (hunk 1 above)
Comment (kathleenlan) [inline README.md]: chunks

File: README.md [hunk 4]
```diff
@@ -142,14 +146,24 @@ Model.reset_column_information
 
 # JavaScript
 * Use CoffeeScript over JavaScript
+* Namespace javascript objects for related chuncks of funcationlity (Components, Engines, etc...)  [example implementation](samples/js_namespace.md)
+* Implement an initializers file applied to any content loaded on the page [see example](samples/initializers.js.md)
```
Comment (kathleenlan) [inline README.md]: This is valid for ES5 and lower.

File: README.md [hunk 5]
```diff
@@ -142,14 +146,24 @@ Model.reset_column_information
...
 * Use a package manager like Bower for installing dependencies? (This item is questionable as some people raised issues with using Bower with Rails. We can discuss further.)
 * Prefer a JavaScript framework over vanilla JavaScript. (This means don't roll your own custom event library or other things that exist out there. But do make sure the company is on board with any new libraries that are used before including them in a project)
 * Use IIFE or Object Literal notation
 * Separate responsibilities of vanilla JavaScript into separate entities.
 * Query DOM elements using `js-` or `data-` attributes instead of CSS classes
 * Do not rely on or store data in the DOM. (This is particularly true from a security stand point. Do not put secure or sensitive information in the DOM. For example do not have a hidden field with a SSN, or that determines whether someone is an admin).
+* Coffeescript file extensions not include `.js` where possible. Prefer `foo.coffee` to `foo.js.coffee.`
+* Always prefix jQuery variables with `$` e.g. `$finder = $('.finder')`
+* For ternary operations in coffeescript, prefer `if foo then bar else baz`
```
Comment (kathleenlan) [inline README.md]: Apparently the ? : syntax has surprising behavior.

File: README.md [hunk 6]
```diff
@@ -142,14 +146,24 @@ Model.reset_column_information
...
+* Use [sweetAlert.js](https://github.com/KMontag42/sweetalert2-rails) for popups (disable animations)
```
Comment (kathleenlan) [inline README.md]: Remove this from style-guide.

File: README.md [hunk 7]
```diff
@@ -142,14 +146,24 @@ Model.reset_column_information
...
+* Prefer JS templating libraries (Handlebars, Mustache) to rolling your own over-complicated JS
```
Comment (kathleenlan) [inline README.md]: Relevant for applications that aren't using ReactJS.

File: samples/scss.md [hunk 8]
```diff
@@ -0,0 +1,37 @@
+#SCSS Whitespace Guidelines
+
+* One line after section comment
+* Two lines before section comment
```
Comment (kathleenlan) [inline samples/scss.md]: Change this to one line before section comment.

File: README.md [hunk 9]
```diff
@@ -142,14 +146,24 @@ Model.reset_column_information
...
 
 # SCSS
+* See [this example](samples/scss.md) for whitespace usage.
+* Use [semantic-ui](https://github.com/doabit/semantic-ui-sass) where possible
```
Comment (kathleenlan) [inline README.md]: Change to `Use agreed-upon css framework where possible`

File: README.md [hunk 10]
```diff
@@ -164,6 +178,11 @@ Model.reset_column_information
 * Use `box-sizing: border-box;` (Whenever browsers support it, this allows you to define the size of actual boxes, rather than the size before padding and borders are added. Generally works for IE8+).
 * Consider the organizational advantages of SMACSS. (We should link to our SMACSS guide here.)
 * Sometimes files should be named `my_styles.css.scss.erb`, sometimes it should be `my_styles.scss.erb` (Will - to add detail)
+* Write out the full hex code in uppercase. e.g. `#AAAAAA` instead of `#aaa`
+* Avoid meta-programming classes
+* Sass file extensions not include `.css` where possible. Prefer `foo.scss` to `foo.css.scss`.
```
Comment (kathleenlan) [inline README.md]: missing `should` in first sentence

File: README.md [hunk 11]
```diff
@@ -164,6 +178,11 @@ Model.reset_column_information
...
+* Follow smacss guidelines for css organization (Base, Layout, Module, State, Theme). Avoid model-specific css.
```
Comment (kathleenlan) [inline README.md]: * Add link to SMACSS guidelines

File: README.md [hunk 12]
```diff
@@ -164,6 +178,11 @@ Model.reset_column_information
 * Use `box-sizing: border-box;` (Whenever browsers support it, this allows you to define the size of actual boxes, rather than the size before padding and borders are added. Generally works for IE8+).
 * Consider the organizational advantages of SMACSS. (We should link to our SMACSS guide here.)
 * Sometimes files should be named `my_styles.css.scss.erb`, sometimes it should be `my_styles.scss.erb` (Will - to add detail)
+* Write out the full hex code in uppercase. e.g. `#AAAAAA` instead of `#aaa`
+* Avoid meta-programming classes
+* Sass file extensions not include `.css` where possible. Prefer `foo.scss` to `foo.css.scss`.
+* Follow smacss guidelines for css organization (Base, Layout, Module, State, Theme). Avoid model-specific css. 
+* Always use a [css reset](https://github.com/markmcconachie/normalize-rails)
```
Comment (kathleenlan) [inline README.md]: * `Always use a css reset. If the css framework being used does not provide one, consider using normalize-rails (link).`

File: README.md [hunk 13]
```diff
@@ -175,6 +194,8 @@ Model.reset_column_information
 * Use of XHTML markup is discouraged e.g. `<br />`
 * Use layout tags (e.g. `<section>`, `<header>`, `<footer>`, `<aside>`) You are not bound to only having one `<header>` or one `<footer>` tag on a page. (Note that you can have only one `<main>` tag however).
 * Put the javascript includes in the bottom of your page, not in the top of the page.
+* Modals should be in a partial suffixed with '_modal.html.erb'
```
Comment (kathleenlan) [inline README.md]: * Change quotes to backticks

File: README.md [hunk 14]
```diff
@@ -175,6 +194,8 @@ Model.reset_column_information
...
+* Double-quote HTML attributes e.g. `<i id="foo"></i>`
```
Comment (kathleenlan) [inline README.md]: * Change to `Double-quote raw HTML attributes`

**Pull request merged**